WMF Framework/EventStreams.swift

```diff
@@ -0,0 +1,26 @@
+import Foundation
+
+extension EPC {
+    /**
+     * Streams are the event stream identifiers that can be utilized with the EventPlatformClientLibrary. They should
+     *  correspond to the `$id` of a schema in
+     * [this repository](https://gerrit.wikimedia.org/g/schemas/event/secondary/).
+     */
+    public enum Stream: String, Codable {
+        case editHistoryCompare = "ios.edit_history_compare"
+    }
+    
+    /**
+     * Schema specifies which schema (and specifically which version of that schema)
+     * a given event conforms to. Analytics schemas can be found in the jsonschema directory of
+     * [secondary repo](https://gerrit.wikimedia.org/g/schemas/event/secondary/).
+     * As an example, if instrumenting client-side error logging, a possible
+     * `$schema` would be `/mediawiki/client/error/1.0.0`. For the most part, the
+     * `$schema` will start with `/analytics`, since there's where
+     * analytics-related schemas are collected.
+     */
+    public enum Schema: String, Codable {
+        case editHistoryCompare = "/analytics/mobile_apps/ios_edit_history_compare/1.0.0"
```

Comment:
I'm a bit apprehensive about this.

The schema is a contract between the event producer and the event ingester. Without this, each piece of instrument declares explicitly which schema and which version of that schema it conforms to by including `$schema` in the event data.

When a schema is updated to have more fields or more allowed values in enum fields and the instrumentation is updated to match, I worry that having the schema+version centralized here, separate from the instrumentation, will make it too easy to forget to update. Of course, if the schema is declared explicitly in the same file as the instrumentation code it's also entirely possible to miss in that case too, but not as easy as if it's different locations.

Also, I really really hope that this isn't taken as me underestimating anybody!

Reply:
That's a good point! It would likely help with maintainability to have the schema and the event structure in the same place. I will push a patch that moves the `schema` definition inside the structure of the event so we can see how it feels.

Reply:
I pushed this change and I agree it helps. Here's what an event looks like after the update:
```swift
    private struct Event: EventInterface {
        static let schema = "/analytics/mobile_apps/ios_edit_history_compare/1.0.0"
        let action: Action
        let primary_language: String
        let is_anon: Bool
    }
```

Reply:
@bearloga would we ever need to share schema identifiers across multiple types of event logs? That's the only potential problem I see with this approach, multiple string literals of the same identifier peppered throughout as we move forward. But I could definitely be misremembering how the backend works and streams vs. schemas.

Reply:
Another potential option would be to keep the `static` definition in the struct but use `EPC.Schema` for the type instead of `String`:
```swift
private struct Event: EventInterface {
        static let schema: EPC.Schema = .editHistoryCompareV1
        let action: Action
        let primary_language: String
        let is_anon: Bool
    }
```

Reply:
Quick recap of MEP to clarify some of this stuff: In the legacy system stream=table=schema. In the modern system schemas and streams have a 1:many relationship. Multiple streams can declare the same schema, and that schema specifies what the incoming the data should look like and how the table in the database looks like. Each stream gets its own table.

When you produce an event to the stream there's 3 conditions which are checked: (1) that the stream exists in the stream config, (2) that the title of the schema in the stream config matches the schema declared by the event, and (3) that the event data conforms to the schema it claims to be conforming to.

I can imagine having a more generic schema that is re-used by multiple instruments that send data to different streams. For example, we can generalize `ios_edit_history_compare` schema by making the `action` field a free-form string instead of an enum like it currently is.

Reply:
Makes sense, thanks for the recap! Given that we might re-use schemas, I pushed the change that restores the central definition of schemas, but keeps the per-event static schema. This way we could re-use them but would also have them declared on the event struct so it's easier to remember it should be updated when changing the event's structure.